.github/workflows/publish.yml

```diff
@@ -2,47 +2,170 @@ name: Publish
 
 on:
   release:
-    types: [ created ]
+    types: [created]
+
+permissions:
+  contents: read
 
 jobs:
```

Comment:
```
name: Publish

on:
  release:
    types: [created]

permissions:
  contents: read

jobs:
  linux:
    runs-on: ${{ matrix.platform.runner }}
    strategy:
      matrix:
        platform:
          - runner: ubuntu-latest
            target: x86_64
          - runner: ubuntu-latest
            target: x86
          - runner: ubuntu-latest
            target: aarch64
          - runner: ubuntu-latest
            target: armv7
          - runner: ubuntu-latest
            target: s390x
          - runner: ubuntu-latest
            target: ppc64le
    steps:
      - <<: *checkout
      - <<: *setup-python
      - <<: *build-wheels
      - <<: *upload-wheels
        with:
          name: wheels-linux-${{ matrix.platform.target }}

  musllinux:
    runs-on: ${{ matrix.platform.runner }}
    strategy:
      matrix:
        platform:
          - runner: ubuntu-latest
            target: x86_64
          - runner: ubuntu-latest
            target: x86
          - runner: ubuntu-latest
            target: aarch64
          - runner: ubuntu-latest
            target: armv7
    steps:
      - <<: *checkout
      - <<: *setup-python
      - <<: *build-wheels
        with:
          manylinux: musllinux_1_2
      - <<: *upload-wheels
        with:
          name: wheels-musllinux-${{ matrix.platform.target }}

  windows:
    runs-on: ${{ matrix.platform.runner }}
    strategy:
      matrix:
        platform:
          - runner: windows-latest
            target: x64
          - runner: windows-latest
            target: x86
    steps:
      - <<: *checkout
      - <<: *setup-python
        with:
          architecture: ${{ matrix.platform.target }}
      - <<: *build-wheels
      - <<: *upload-wheels
        with:
          name: wheels-windows-${{ matrix.platform.target }}

  macos:
    runs-on: ${{ matrix.platform.runner }}
    strategy:
      matrix:
        platform:
          - runner: macos-12
            target: x86_64
          - runner: macos-14
            target: aarch64
    steps:
      - <<: *checkout
      - <<: *setup-python
      - <<: *build-wheels
      - <<: *upload-wheels
        with:
          name: wheels-macos-${{ matrix.platform.target }}

  sdist:
    runs-on: ubuntu-latest
    steps:
      - <<: *checkout
      - name: Build sdist
        uses: PyO3/maturin-action@v1
        with:
          command: sdist
          args: --out dist
      - <<: *upload-wheels
        with:
          name: wheels-sdist

  release:
    name: Release
    runs-on: ubuntu-latest
    if: ${{ startsWith(github.ref, 'refs/tags/') || github.event_name == 'workflow_dispatch' }}
    needs: [linux, musllinux, windows, macos, sdist]
    permissions:
      id-token: write
      contents: write
      attestations: write
    steps:
      - uses: actions/download-artifact@v4
      - name: Generate artifact attestation
        uses: actions/attest-build-provenance@v1
        with:
          subject-path: "wheels-*/*"
      - name: Publish to PyPI
        if: "startsWith(github.ref, 'refs/tags/')"
        uses: PyO3/maturin-action@v1
        env:
          MATURIN_USERNAME: ${{ secrets.PYPI_USERNAME }}
          MATURIN_PASSWORD: ${{ secrets.PYPI_PASSWORD }}
        with:
          command: upload
          args: --non-interactive --skip-existing wheels-*/*

# YAML Anchors
checkout: &checkout
  uses: actions/checkout@v4

setup-python: &setup-python
  uses: actions/setup-python@v5
  with:
    python-version: 3.x

build-wheels: &build-wheels
  name: Build wheels
  uses: PyO3/maturin-action@v1
  with:
    target: ${{ matrix.platform.target }}
    args: --release --out dist --find-interpreter
    sccache: "true"
    manylinux: auto

upload-wheels: &upload-wheels
  name: Upload wheels
  uses: actions/upload-artifact@v4
  with:
    path: dist
```

Reply:
@SaulTigh I've asked chatGpt to cut repetitive parts, you could utilise it if u like, altough it did not cut the runners part

Reply:
@wRLSS thanks, but i would avoid using snippet you provided because this file is mostly generated by tooling (with few tweaks from my side), this will allow file replacement as easy as just call `maturin generate-ci github` :)